keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,622 @@
+ 
+```
+ 
+                    Ingress: []networkingv1.NetworkPolicyIngressRule{{
+                        From: []networkingv1.NetworkPolicyPeer{
+                            {
+                                // TODO add these composably, so that can be disambiguated from combo networkpolicypeer
+                                PodSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "pod-name": "client-b",
+                                    },
+                                },
+                            },
+                            {
+                                NamespaceSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "ns-name": nsBName,
+                                    },
+                                },
+                            },
+                        },
+                    }},
+```
+ 
+The AND test is obviously more selective, although it is tricky to tell from the struct that
+it has been correctly written to be different from the OR test.
+ 
+```
+                    Ingress: []networkingv1.NetworkPolicyIngressRule{{
+                        From: []networkingv1.NetworkPolicyPeer{
+                            {
+                                PodSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "pod-name": "client-b",
+                                    },
+                                },
+                                // because we lack {,} , this is a single Peer selecting
+                                // pods, from namespaces selected by namespace selector.
+                                // This is difficult to verify for correctness at a
+                                // glance, due to the verbosity of the struct.
+                                NamespaceSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "ns-name": nsBName,
+                                    },
+                                },
+                            },
+                        },
+```
+
+In contrast, we can express the same and test using the API (including its entire connectivity matrix) in this proposal as follows:
+
+```
+	builder := &NetworkPolicySpecBuilder{}
+	builder = builder.SetName("myns", "allow-podb-in-nsb").SetPodSelector(map[string]string{"pod": "b"})
+	builder.SetTypeIngress()
+	builder.AddIngress(nil, &p80, nil, nil, map[string]string{"pod-name": "b"}, nil, map[string]string{"ns-name": "b"}, nil)
+	policy := builder.Get()
+	reachability := NewReachability(allPods, false)
+	reachability.ExpectAllIngress(Pod("myns/b"), false)
+	reachability.Expect(Pod("b/b"), Pod("myns/b"), true)
+	reachability.Expect(Pod("myns/b"), Pod("myns/b"), true)
```

Comment:
OTOH, making a system for easily defining a bunch of pod tests does seem useful

Reply:
yup :) we'll upstream this to sig-test as well, something @bowei suggested as well